pkg/awsutils/awsutils.go

```diff
@@ -474,6 +474,11 @@ func (cache *EC2InstanceMetadataCache) RefreshSGIDs(mac string) error {
 			_, err = cache.ec2SVC.ModifyNetworkInterfaceAttributeWithContext(context.Background(), attributeInput)
 			awsAPILatency.WithLabelValues("ModifyNetworkInterfaceAttribute", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
 			if err != nil {
+				if aerr, ok := err.(awserr.Error); ok {
+					if aerr.Code() == "InvalidNetworkInterfaceID.NotFound" {
```

Comment:
Curious, if this check is enough for a leaked ENI case? (active but not attached to any instance). Interface ID is still valid but the ENI is probably no longer attached to the Instance (or any instance in the cluster). I mean before the IMDS is updated, it is possible we will actually go and update the SGs on this instance. Should we check if the ENI is attached to an instance (from describeENI o/p) before we call Modify on that ENI.

Maybe we can count on the leaked eni routine to delete that ENI eventually(?)...VPC CNI actually goes ahead and updates the SGs on this ENI even though it is no longer tied to this cluster but the leaked ENI routine will eventually delete it. Is that the expectation here?

Reply:
Yeah, we might have to check `attachment.instance-id - The ID of the instance to which the network interface is attached.` Similar check was done here - https://github.com/aws/amazon-vpc-cni-k8s/blob/9db2ae62ecd0cb56f7fc20b80427fa6ff4e17a42/pkg/awsutils/awsutils.go#L1007 . Yeah the design looks like the leaked ENI will eventually delete the leaked ENI.